Add illink analyzer support for field/property initializers

sbomer opened the pull request:
The dataflow analyzer was exiting early for cases where the owning symbol was not an `IMethodSymbol`. This meant we weren't running dataflow analysis for field and property initializers.

This fixes it by allowing through cases where the owning symbol is not an `IMethodSymbol`, and adding testcases to validate that we don't hit asserts in the code paths that only light up for methods.

sbomer committed:
Fix missing dataflow warnings on field initializers

sbomer committed:
Adjust asserts, add comments

sbomer committed:
Fix test for ILLink/ILCompiler

vitek-karas commented on src/tools/illink/test/Mono.Linker.Tests/TestCasesRunner/ResultChecker.cs:
```diff
@@ -880,7 +880,7 @@ void VerifyLoggedMessages (AssemblyDefinition original, TrimmingTestLogger logge
 											break;
 										}
 										if (memberDefinition.Name == ".ctor" &&
```
Nit: This should probably also have `actualName.StartsWith (expectedMember.DeclaringType.FullName)` just like the static `.cctor` above.

vitek-karas commented on src/tools/illink/src/ILLink.RoslynAnalyzer/DataFlow/LocalDataFlowVisitor.cs:
```diff
@@ -504,7 +505,7 @@ public override TValue VisitObjectCreation (IObjectCreationOperation operation,
 		public override TValue VisitFlowAnonymousFunction (IFlowAnonymousFunctionOperation operation, LocalDataFlowState<TValue, TValueLattice> state)
 		{
-			Debug.Assert (operation.Symbol.ContainingSymbol is IMethodSymbol);
+			Debug.Assert (operation.Symbol.ContainingSymbol is IMethodSymbol or IFieldSymbol);
```
Why not property as well? What does this translate to:
```csharp
class Helper
{
    public Action<int> DoIt => (x) => Console.WriteLine(x);
}
```

I guess it could be a getter with the body as well...

sbomer committed:
PR feedback

- Add comments about assert, with testcases
- Ensure cctor warning origins match expected type name
